[REG-936] Co-op archer code

vontell opened the pull request:
Adds a bot that can be used for some basic co-op. It follows the player, attacks enemies, and holds the switch for the player.

NOTE: Maybe we don't merge this in and keep it as a separate branch? I don't want to override the existing ability cycling bot. Actually, I will simply move the two bots into different folders. @reviewers id say review this now since a lot more changes that are just moving things may appear once I do that - I'll wait for approvals and then do that change. Also depends on this PR: https://github.com/Regression-Games/RGBossRoom/pull/12

https://www.loom.com/share/53aa8ab4b0b946cf91a138433210e94a

RG-nAmKcAz commented on bossroom.js:
```diff
@@ -40,7 +48,15 @@ export const BossRoomBot = {
         actionQueue.queue("PerformSkill", input)
+    },
+
+    moveTowards: (target, range, actionQueue) => {
```
move towards and follow object don't mean the same thing to me, please rename this to followTarget or followObject

vontell replied:
Fixed!

RG-nAmKcAz commented on index.js:
```diff
@@ -1,12 +1,8 @@
 export function configureBot(characterType) {
-  console.log(`Unity bot configureBot function called, charType: ${charType} - characterType: ${characterType}`);
-  charType = CharInfo.type.indexOf(characterType);
+  console.log(`Unity bot configureBot function called, charType: ${getCharacterType()} - characterType: ${characterType}`);
```
this breaks the server random assignment case , please revert this change

vontell replied:
For the co-op bot I wanted to only use the archer, and then was going to have the ability bot keep this functionality. I was a little confused by this code actually, if the getCharacterType function is configuring the bot type, what is this setting exactly?

RG-nAmKcAz replied:
charType field is randomly assigned when the bot code instantiates

so getCharacterType() will return that value until the handshake completes (unless you override it)

once the handshake completes

 ... if you supplied a value for characterType, you will get that same value back and charType will be set to that same value

 ... if you didn't supply a value for characterType, you get some random value back which will then be set as your charType

vontell replied:
Got it, thanks!

RG-nAmKcAz commented on bossroom.js:
```diff
@@ -32,6 +32,14 @@ export const BossRoomBot = {
     },
 
+    getDoorSwitch: (tickInfo) => {
+        return RGBot.getEntitiesOfType(tickInfo, "FloorSwitch")[0];
```
There are 2 door switches in the game, did you only add the sensor to the one in the first room ?  Otherwise this is non-deterministic whether it will work reliably as it may pick the switch behind the door.

vontell replied:
Yeah I only added it to the first one for this demonstration.

RG-nAmKcAz commented on index.js:
```diff
@@ -16,15 +12,41 @@ let lastEnemyId = -1;
+ *  - If the bot is standing on the switch, do nothing
+ *  - If the bot is not near the player, move within range of the player
+ *  - If an enemy is within a certain distance of a player, attack that enemy
+ *  - If the switch is within a range of 5 units from the bot, move onto the switch
```
Commented distances don't match actual code, keep in mind these are distance^2 values (looks like you already considered that :D )

vontell replied:
Will fix!

vontell replied:
Fixed! Yeah this is why I put units, which I am thinking in my head as "squared SOMETHING" but idk what they even call the distance in Unity

vontell commented:
@RG-nAmKcAz I've added back the ability code and moved that and this new code into separate folders - none of the code in the ability bot folder is new.

Went and fixed your suggestions as well. Thanks! Quick question though - what happens if two bots try to select the same character type?

vontell commented:
@RG-nAmKcAz I've added back the ability code and moved that and this new code into separate folders - none of the code in the ability bot folder is new.

Went and fixed your suggestions as well. Thanks! Quick question though - what happens if two bots try to select the same character type?

vontell commented:
@RG-nAmKcAz I've added back the ability code and moved that and this new code into separate folders - none of the code in the ability bot folder is new.

Went and fixed your suggestions as well. Thanks!

vontell commented:
Also, here is an updated video with 2 ability bots and 1 co-op bot

https://www.loom.com/share/2199ea546f7549aa901e6bd66f9931f8

RG-nAmKcAz reviewed (changes requested):
These changes mean that OOB , our bots in the database are now broken because the path changed.

Maybe update system user seeder to just handle this in the unity branch by creating/updating the user with that name and making sure it has the right bots ?

RG-nAmKcAz reviewed (changes requested):
These changes mean that OOB , our bots in the database are now broken because the path changed.

Maybe update system user seeder to just handle this in the unity branch by creating/updating the user with that name and making sure it has the right bots ?

RG-nAmKcAz reviewed (changes requested):
These changes mean that OOB , our bots in the database are now broken because the path changed.

Maybe update system user seeder to just handle this in the unity branch by creating/updating the user with that name and making sure it has the right bots ?

vontell commented:
> These changes mean that OOB , our bots in the database are now broken because the path changed.
> 
> Maybe update system user seeder to just handle this in the unity branch by creating/updating the user with that name and making sure it has the right bots ?

Because of how the SQL script is written, the script will fail no matter what if you run it again because it uses INSERTS, so even the user creation part fails due to conflicting primary keys. Would you prefer I go and fix all of those to also do updates, or just fix that one line and assume whoever runs it is running from the beginning?
In other words, even if I rewrite the part im working on to be correct now, you'll have to skip the first few lines if you want to update

vontell commented:
@RG-nAmKcAz that PR has been added here: https://github.com/Regression-Games/RegressionGames/pull/298

abeizer commented on coopbot/index.js:
```diff
@@ -0,0 +1,140 @@
+/**
+ * Selects an ability for this character, and queues that action.
+ */
+function selectAbility(playerId, tickInfo, mostRecentMatchInfo, actionQueue) {
```
It doesn't look like we're using this?
